### PR TITLE
[fixing cuda launch config failure on UpSampleNearest]

### LINKED ISSUE #22526: RuntimeError: CUDA error: invalid configuration argument
## Issue description

when it use  torch._C._nn.upsample_nearest2d(input, _output_size(2)), it's comes with 

RuntimeError: CUDA error: invalid configuration argument. 

The problem is only occur in torch-nightly( 1.2.0.dev20190702).

## Error log

File "/home/xdjf/XXXX/XXXXX/XXXXXXXX.py", line 74, in _upsample_add
    return F.interpolate(x, size=(H,W), mode='nearest') + y
File "/home/XXXX/.conda/envs/py36torch120/lib/python3.6/site-packages/torch/nn/functional.py", line 2545, in interpolate
    return torch._C._nn.upsample_nearest2d(input, _output_size(2))
RuntimeError: CUDA error: invalid configuration argument

**You can get the script and run it with:**
torch-nightly( 1.2.0.dev20190702)
```
import torch
import torch.nn.functional as F

x=torch.rand(4,256,1,1).cuda()
y=torch.rand(4,256,3,3).cuda()

F.interpolate(x, size=(3,3), mode='nearest') + y
```

- PyTorch:
- conda install pytorch-nightly cudatoolkit=10.0 -c pytorch
- OS:Ubuntu16.04 LTS
- PyTorch version: torch-nightly( 1.2.0.dev20190702)
- Python version:3.6
- CUDA/cuDNN version:9.0 or 10.0
- GPU models and configuration: GTX 1080ti * 4

### PR DESCRIPTION
**Pull request by jjsjann123**
This is to fix #22526 

Adding limitation on launch config for grid sizes as well, previous code is asking to launch blocks more than what's supported by the hardware;
Test added in test_cuda;

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 1]
```diff
@@ -160,9 +161,12 @@ static void upsample_nearest2d_out_cuda_template(
       maxThreadsDim[2], std::min<int>(nc, max_threads / block_x / block_y));
   const dim3 block(block_x, block_y, block_z);
 
-  int grid_x = cuda::ATenCeilDiv(output_width, block_x);
-  int grid_y = cuda::ATenCeilDiv(output_height, block_y);
-  int grid_z = cuda::ATenCeilDiv(nc, block_z * 4);
+  int grid_x = std::min<int>(
+      maxGridSize[0], cuda::ATenCeilDiv(output_width, block_x));
+  int grid_y = std::min<int>(
+      maxGridSize[1], cuda::ATenCeilDiv(output_height, block_y));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Granted, your grid_y and especially grid_x are very unlikely to be more than maxGridSize, but if they ever are, you don't have a loop in the kernel to handle that, so better error out. 
Also, if your grid_x is ever more than maxGridSize[0] (which is 2**31), you'll overflow in your offset computation which uses `int` type.

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 2]
```diff
@@ -233,7 +237,9 @@ static void upsample_nearest2d_backward_out_cuda_template(
   unsigned int n = grad_input.numel() / nbatch;
   dim3 bdim{std::min<unsigned int>(
       at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS)};
-  dim3 gdim{cuda::ATenCeilDiv(n, bdim.x)};
+  dim3 gdim{std::min<unsigned int>(
+      at::cuda::getCurrentDeviceProperties()->maxGridSize[0],
+      cuda::ATenCeilDiv(n, bdim.x))};
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: same here, you are not handling it correctly in the kernel, so don't truncate grid dim here just to silently produce wrong results

**Review by ngimel — CHANGES_REQUESTED** [review]
Please make sure that if you are limiting the grid dimensions, you are handling it correctly in the kernel.  I've flagged 2 places where you are not doing that, there are likely others. Also, you've added a single test that's not testing everything you changed, and it is also not testing correctness, just that you are not erroring out. That's the wrong kind of test to find silent wrong answers.

**Comment by jjsjann123** [comment]
@ngimel good catch!

Let me take a better look tomorrow to make sure it errors out in cases where the kernel would otherwise produce wrong results.

**Comment by soumith** [comment]
okay I'm reverting the PR, sorry. I'll wait for a new PR with changes that Natalia asked for.

**Comment by zhanwenchen** [comment]
Any update?

**Comment by ngimel** [comment]
#29016

**Pull request closed without merging**